core/src/encode.rs

```diff
@@ -0,0 +1,171 @@
+use thiserror::Error;
+
+#[derive(Deserialize, Serialize)]
+enum Encoded {
+    Bincode(Vec<u8>),
+    Deflate(Vec<u8>),
+    Gzip(Vec<u8>),
+    Zlib(Vec<u8>),
+    Zstd(Vec<u8>),
+}
+
+#[derive(Clone, Copy)]
+pub enum Options {
+    Bincode,
+    Deflate(Compression),
+    Gzip(Compression),
+    Zlib(Compression),
+    Zstd { level: i32 },
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("io error")]
+    IoError(#[from] std::io::Error),
+    #[error("(de)serialization error")]
+    SerializationError(#[from] bincode::Error),
+}
+
+impl Encoded {
+    fn len(&self) -> usize {
+        match self {
+            Encoded::Bincode(bytes) => bytes.len(),
+            Encoded::Deflate(bytes) => bytes.len(),
+            Encoded::Gzip(bytes) => bytes.len(),
+            Encoded::Zlib(bytes) => bytes.len(),
+            Encoded::Zstd(bytes) => bytes.len(),
+        }
+    }
+}
+
+fn encode_zstd(data: &[u8], level: i32) -> std::io::Result<Vec<u8>> {
+    let mut buffer = Vec::with_capacity(data.len() * 2);
+    let mut encoder = zstd::stream::write::Encoder::new(&mut buffer, level)?;
+    encoder.write_all(data)?;
+    encoder.finish()?;
+    Ok(buffer)
+}
+
+macro_rules! encode {
+    ($encoder:ident, $bytes:ident, $compression:ident) => {{
+        fn encode(data: &[u8], compression: Compression) -> std::io::Result<Vec<u8>> {
+            let mut buffer = Vec::with_capacity(data.len() * 2);
+            let mut encoder = $encoder::new(&mut buffer, compression);
+            encoder.write_all(data)?;
+            encoder.finish()?;
+            Ok(buffer)
+        }
+        encode(&$bytes[..], $compression).ok()
+    }};
+}
+
+pub fn encode<T: Serialize>(obj: &T, options: Options) -> bincode::Result<Vec<u8>> {
+    let bytes = bincode::options().serialize(obj)?;
+    let encoded = match options {
+        Options::Bincode => None,
+        Options::Deflate(opts) => encode!(DeflateEncoder, bytes, opts).map(Encoded::Deflate),
+        Options::Gzip(opts) => encode!(GzEncoder, bytes, opts).map(Encoded::Gzip),
+        Options::Zlib(opts) => encode!(ZlibEncoder, bytes, opts).map(Encoded::Zlib),
+        Options::Zstd { level } => encode_zstd(&bytes, level).map(Encoded::Zstd).ok(),
+    };
+    let encoded = match encoded {
+        Some(encoded) if encoded.len() < bytes.len() => encoded,
+        _ => Encoded::Bincode(bytes),
+    };
+    bincode::options().serialize(&encoded)
+}
+
+pub fn decode<T: DeserializeOwned>(bytes: &[u8]) -> Result<T, Error> {
+    let encoded = bincode::options().deserialize_from(bytes)?;
+    let decoder: Box<dyn Read> = match &encoded {
+        Encoded::Bincode(bytes) => Box::new(&bytes[..]),
+        Encoded::Deflate(bytes) => Box::new(DeflateDecoder::new(&bytes[..])),
+        Encoded::Gzip(bytes) => Box::new(GzDecoder::new(&bytes[..])),
+        Encoded::Zlib(bytes) => Box::new(ZlibDecoder::new(&bytes[..])),
+        Encoded::Zstd(bytes) => Box::new(zstd::stream::read::Decoder::new(&bytes[..])?),
```

Comment:
Are these decoders hardened against zip-bomb-like behavior?

Reply:
I raised that question to the upstream package; and meanwhile, added a `limit` here for number of bytes decoded.